cluster/kubernetes/sync.go

```diff
@@ -2,31 +2,293 @@ package kubernetes
 
 import (
 	"bytes"
+	"crypto/sha1"
+	"encoding/hex"
 	"fmt"
 	"io"
 	"os/exec"
 	"sort"
 	"strings"
 	"time"
 
-	rest "k8s.io/client-go/rest"
-
 	"github.com/go-kit/kit/log"
+	"github.com/imdario/mergo"
 	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	rest "k8s.io/client-go/rest"
+
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/cluster"
+	kresource "github.com/weaveworks/flux/cluster/kubernetes/resource"
+	"github.com/weaveworks/flux/policy"
+	"github.com/weaveworks/flux/resource"
+)
+
+const (
+	syncSetLabel       = kresource.PolicyPrefix + "sync-set"
+	checksumAnnotation = kresource.PolicyPrefix + "sync-checksum"
+
+	// The namespace to presume if something doesn't have one, and we
+	// haven't been told what to use as a fallback. This is what
+	// `kubectl` uses when there's no config setting the fallback
+	// namespace.
+	DefaultFallbackNamespace = "default"
 )
 
+// Sync takes a definition of what should be running in the cluster,
+// and attempts to make the cluster conform. An error return does not
+// necessarily indicate complete failure; some resources may succeed
+// in being synced, and some may fail (for example, they may be
+// malformed).
+func (c *Cluster) Sync(spec cluster.SyncSet) error {
+	logger := log.With(c.logger, "method", "Sync")
+
+	// Keep track of the checksum of each resource, so we can compare
+	// them during garbage collection.
+	checksums := map[string]string{}
+
+	// NB we get all resources, since we care about leaving unsynced,
+	// _ignored_ resources alone.
+	clusterResources, err := c.getResourcesBySelector("")
```

Comment:
For follow-up at some point:

Here we want to know about resources corresponding to manifests we have in the git repo, so we can see if they are ignored.

We can do better than asking for all resources. For a start, the manifests in the cluster will only be a subset of the API resources from the cluster, so we need not ask for all of those. Moreover, it's exactly those in the manifests, that aren't already annotated as being ignored, that we need to look up; so we only need to ask for specific namespaces, and in those, specific resources.

Only requesting the API resources needed would come at little cost (a pass through the manifests, or a pull-through cache of some sort). That seems like a good trade.

Restricting the namespaces searched again seems like a good trade, since that reduces the number of requests with little cost.

Asking for the specific resources would probably be a step too far, though -- in general, most resources won't be ignored, and requesting lists of resources is more efficient than asking for individual resources.

Reply:
@squaremo Mind adding a comment/issue covering this?